android/src/org/mozilla/firefox/vpn/InAppPurchase.kt

```diff
@@ -0,0 +1,350 @@
+            .build()
+        val billingResult = billingClient.launchBillingFlow(activity, billingParams)
+        if (billingResult.responseCode != BillingClient.BillingResponseCode.OK) {
+            onSubscriptionFailed(billingResultToJson(billingResult, "initiatePurchase"))
+        }
+    }
+
+    fun initiateAcknowledge(purchaseToken: String) {
+        val acknowledgePurchaseParams = AcknowledgePurchaseParams.newBuilder()
+            .setPurchaseToken(purchaseToken)
+            .build()
+        billingClient.acknowledgePurchase(acknowledgePurchaseParams, this)
+    }
+
+    /**
+     * Override functions for Billing Library listeners
+     */
+
+    override fun onBillingSetupFinished(billingResult: BillingResult) {
+        if (billingResult.responseCode != BillingClient.BillingResponseCode.OK) {
+            onBillingNotAvailable(billingResultToJson(billingResult, "onBillingSetupFinished"))
+        } else {
+            querySkuAndPurchases()
+        }
+    }
+
+    override fun onBillingServiceDisconnected() {
+        Log.i(TAG, "Billing Service Disconnected")
+        skusWithSkuDetails.clear()
+        skusWithMonthCount.clear()
+        skusWithPurchase.clear()
+        onBillingNotAvailable(
+            Json.encodeToString(
+                BillingResultData(
+                    code = -99,
+                    message = "Billing Service Disconnected"
+                )
+            )
+        )
+    }
+
+    override fun onSkuDetailsResponse(
+        billingResult: BillingResult,
+        skuDetailsList: MutableList<SkuDetails>?
+    ) {
+        if (billingResult.responseCode != BillingClient.BillingResponseCode.OK) {
+            onSkuDetailsFailed(billingResultToJson(billingResult, "onSkuDetailsResponse"))
+            return
+        }
+        if (skuDetailsList == null) {
+            onSkuDetailsFailed(
+                Json.encodeToString(
+                    BillingResultData(
+                        code = -97,
+                        message = "No sku details returned"
+                    )
+                )
+            )
+            return
+        }
+        val googleProducts = GooglePlaySubscriptions(products = arrayListOf())
+        for (details in skuDetailsList) {
+            val parsedDetails = skuDetailsToGooglePlaySubscriptionInfo(details)
+            if (parsedDetails != null) {
+                googleProducts.products.add(parsedDetails)
+            }
+        }
+        val googleProductsJson = Json.encodeToString(googleProducts)
+        Log.d(TAG, "Sending $googleProductsJson")
+        onSkuDetailsReceived(googleProductsJson)
+    }
+
+    override fun onQueryPurchasesResponse(
+        billingResult: BillingResult,
+        purchases: MutableList<Purchase>
+    ) {
+        if (billingResult.responseCode == BillingClient.BillingResponseCode.OK) {
+            processPurchases(purchases)
+        }
+        // TODO - Could try re-querying up to n times before declaring failure
```

Comment:
Does the documentation say to do it? If not, I guess we are fine with what you have implemented.

Reply:
There's no documentation for this particular case. I'll just add a logging statement in the off chance we start hitting it at least there'll be a breadcrumb.